resources-generator/src/main/kotlin/dev/icerock/gradle/generator/StringsGenerator.kt

```diff
@@ -13,9 +13,36 @@ import dev.icerock.gradle.generator.jvm.JvmStringsGenerator
 import dev.icerock.gradle.utils.removeLineWraps
 import org.gradle.api.file.FileTree
 import java.io.File
+import java.util.Locale as JvmLocale
 import javax.xml.parsers.DocumentBuilderFactory
 
-typealias LanguageType = String
+sealed interface LanguageType {
+
+    object Base : LanguageType
+
+    class Locale(languageTag: String) : LanguageType {
+
+        val locale: JvmLocale = JvmLocale.forLanguageTag(languageTag)
+
+        override fun hashCode(): Int = locale.hashCode()
+        override fun toString(): String = locale.toLanguageTag()
+        override fun equals(other: Any?): Boolean {
+            return other is Locale && other.locale == locale
+        }
+    }
+
+    companion object {
+
+        private const val BASE = "base"
+
+        fun fromFileName(fileName: String): LanguageType = when (fileName) {
+            BASE -> Base
+            else -> Locale(fileName.replace("-r", "-"))
```

Comment:
Implementation note: 

The file we specify this in should look like `en-rUS`. This takes inspiration from Android, but I wasn't happy to do this. I was running into weird issues with Gradle's `FileTree` where simply specifying `en-US` was producing duplicate (?) files. 

We should investigate this farther to see if this disambiguation is necessary.

Reply:
i not see in android docs about r prefix...
https://developer.android.com/guide/topics/resources/multilingual-support
maybe it's reason why tests failed?

Reply:
Going to look more into this today.

Reply:
I think the failures are because of the following from the documentation: 

> Prior to Android 7.0, Android could not always successfully match app and system locales.
>
> For example, assume that you have the following situation:
>
>    Your app's default language is en_US (US English), and it also has Spanish strings localized in es_ES resource files.
>    A device is set to es_MX
>
> When your Java code refers to strings, the system would load strings from the default (en_US) resource file, even if the app has Spanish resources localized under es_ES. This is because when the system cannot find an exact match, it continues to look for resources by stripping the country code off the locale. Finally, if no match is found, the system falls back to the default, which is en_US. 
> 
> ... 
> 
> Android 7.0 (API level 24) brings more robust resource resolution, and finds better fallbacks automatically. However, to speed up resolution and improve maintainability, you should store resources in the most common parent dialect. For example, if you were storing Spanish resources in the values-es-rUS directory before, move them into the values-b+es+419 directory, which contains Latin American Spanish. Similarly, if you have resource strings in a directory named values-en-rGB, rename the directory to values-b+en+001 (International English), because the most common parent for en-GB strings is en-001. The following examples explain why these practices improve performance and reliability of resource resolution.

This is making me feel that `en` resources should generally be disjoint sets from `en-XX`. For example: 

```xml
    < -- en-US -->
    <string name="test.simple">test US</string>

    <-- en-GB -->
    <string name="test.simple">test US</string>

    <-- en (Base) -->
    <string name="test.simple">test</string>
```

When localizing for `en`, I don't believe that Android will attempt to match the `Base` value first. It will instead try to match either `en-GB` or `en-US`, fail, and then strip off the dialects from the resource lookup. Then it will find a partial match in either `en-GB` or `en-US` and use that. 

The Android documentation points this out. I think the _correct_ course of action for end consumers of this library is to make sure keys that are in regional dialect resources (eg. `test.simple`) do not appear in the base language `en` resource folder. Therefore, these language dialects should really only be used for things which are truly different between different regions. 

What are your thoughts?

Reply:
I'm second guessing my thought again. This test. 

```kt
    @Test fun checkNonExistentRegionalDialect() = runTest {
        StringDesc.localeType = StringDesc.LocaleType.Custom("en-CA")
        val rawString: String = MR.strings.test_dialect.desc().getString()
        assertEquals(
            expected = "Base",
            actual = rawString
        )
    }
```

Fails with the following: 

```
expected:<[Base]> but was:<[UK]>
Expected :Base
Actual   :UK
```

I _think_ Android is stripping the `-CA` from the desired dialect and then selecting _any_ language tag which partially matches.

Reply:
Adding a dedicated `en` folder fixed this behavior. This is quite confusing. I'm not sure how we'd smooth this quirk over between platforms (or at least how we'd document it).

Reply:
I think I've figured it out. On Android, default resources don't have a language affinity and are only treated as defaults. Therefore, my above confusion stems from the need to essentially have a duplicate resource set between what my base would be `en` and `base`. 

I think this is a non-issue.